src/main/java/org/jabref/logic/exporter/AtomicFileOutputStream.java

```diff
@@ -0,0 +1,197 @@
+    }
+
+    /**
+     * Returns the path of the backup copy of the original file (may not exist)
+     */
+    public Path getBackup() {
+        return backupFile;
+    }
+
+    /**
+     * Override for performance reasons.
+     */
+    @Override
+    public void write(byte b[], int off, int len) throws IOException {
+        try {
+            out.write(b, off, len);
+        } catch (IOException exception) {
+            cleanup();
+            throw exception;
+        }
+    }
+
+    /**
+     * Closes the write process to the temporary file but does not commit to the target file.
+     */
+    public void abort() throws IOException {
+        try {
+            super.close();
+            Files.deleteIfExists(temporaryFile);
+            Files.deleteIfExists(backupFile);
+        } catch (IOException exception) {
+            LOGGER.debug("Unable to abort writing to file " + temporaryFile, exception);
+        }
+    }
+
+    private void cleanup() {
+        try {
+            Files.deleteIfExists(temporaryFile);
+        } catch (IOException exception) {
+            LOGGER.debug("Unable to delete file " + temporaryFile, exception);
+        }
+    }
+
+    // perform the final operations to move the temporary file to its final destination
+    @Override
+    public void close() throws IOException {
+        try {
+            try {
+                // Make sure we have written everything to the temporary file
+                flush();
+                ((FileOutputStream) out).getFD().sync();
+            } catch (IOException exception) {
+                // Try to close nonetheless
+                super.close();
+                throw exception;
+            }
+            super.close();
+
+            // We successfully wrote everything to the temporary file, lets copy it to the correct place
+            // First, make backup of original file and try to save file permissions to restore them later (by default: 664)
+            Set<PosixFilePermission> oldFilePermissions = EnumSet.of(PosixFilePermission.OWNER_READ,
+                    PosixFilePermission.OWNER_WRITE,
+                    PosixFilePermission.GROUP_READ,
+                    PosixFilePermission.GROUP_WRITE,
+                    PosixFilePermission.OTHERS_READ);
+            if (Files.exists(targetFile)) {
+                Files.copy(targetFile, backupFile, StandardCopyOption.REPLACE_EXISTING);
+                if (FileUtil.IS_POSIX_COMPILANT) {
+                    try {
+                        oldFilePermissions = Files.getPosixFilePermissions(targetFile);
+                    } catch (IOException exception) {
+                        LOGGER.warn("Error getting file permissions for file {}.", targetFile, exception);
+                    }
+                }
+            }
+
+            // Move temporary file (replace original if it exists)
+            Files.move(temporaryFile, targetFile, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
```

Comment:
We have a move in FileUtil. And you better check if the move was successufl

Reply:
Yes, good point - I don't understand why we have a own move method that actually looks more like a "read and rewrite content"-method. Moreover, the move is successful if there is no thrown exception (I assume).